nilearn/glm/first_level/first_level.py

```diff
@@ -124,26 +155,60 @@ def run_glm(Y, X, noise_model='ar1', bins=100, n_jobs=1, verbose=0):
     # Create the model
     ols_result = OLSModel(X).fit(Y)
 
-    if noise_model == 'ar1':
-        # compute and discretize the AR1 coefs
-        ar1 = (
-            (ols_result.residuals[1:]
-             * ols_result.residuals[:-1]).sum(axis=0)
-            / (ols_result.residuals ** 2).sum(axis=0)
-        )
+    if noise_model[:2] == 'ar':
+
+        err_msg = ('AR order must be a positive integer specified as arN, '
+                   'where N is an integer. E.g. ar3. '
+                   'You provied {0}.'.format(noise_model))
+        try:
+            ar_order = int(noise_model[2:])
+        except ValueError:
+            raise ValueError(err_msg)
+
+        # compute the AR coeficents
+        ar_coef_ = [_yule_walker(ols_result.residuals[:, res],
+                                 ar_order)
+                    for res in range(ols_result.residuals.shape[1])]
+        ar_coef_ = np.array(ar_coef_)
         del ols_result
-        ar1 = (ar1 * bins).astype(np.int) * 1. / bins
-        # Fit the AR model acccording to current AR(1) estimates
+        if len(ar_coef_[0]) == 1:
+            ar_coef_ = ar_coef_.ravel()
+
+        # Either bin the AR1 coefs or cluster ARN coefs
+        if ar_order is 1:
+            for idx in range(len(ar_coef_)):
+                ar_coef_[idx] = (ar_coef_[idx] * bins).astype(int) \
+                                * 1. / bins
+            labels = np.array([np.str(val) for val in ar_coef_])
+        else:  # AR(N) case
+            n_clusters = np.min([bins, Y.shape[1]])
+            kmeans = KMeans(n_clusters=n_clusters).fit(ar_coef_)
+            ar_coef_ = kmeans.cluster_centers_[kmeans.labels_]
+
+            # Create a set of rounded values for the labels
+            cluster_labels = kmeans.cluster_centers_.copy()
+            for idx in range(len(cluster_labels)):
+                cluster_labels[idx] = (cluster_labels[idx] * 100).\
+                                          astype(int) * 1. / 100
+
+            cluster_labels = np.array([np.str('_'.join([str(v) for v in val]))
+                 for val in cluster_labels])
```

Comment:
can you use soemthing simpler using np.round ? eg.

```
In [13]: order = 3

In [14]: a = np.random.randn(order)

In [15]: '_'.join(map(str, np.round(a, 2)))
Out[15]: '1.12_-1.17_-1.49'
```

Reply:
OOhhh that is nicer. I think I prefer this approach to the current commit version with too similar code. Let me know what you prefer.